// @pcd/pod/dist/cjs/src/podTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.POD_NULL_HASH = exports.PODNull = exports.POD_DATE_MAX = exports.POD_DATE_MIN = exports.PODEdDSAPublicKeyValue = exports.POD_INT_MAX = exports.POD_INT_MIN = exports.POD_CRYPTOGRAPHIC_MAX = exports.POD_CRYPTOGRAPHIC_MIN = exports.POD_STRING_TYPE_REGEX = exports.EDDSA_PUBKEY_TYPE_STRING = exports.POD_NAME_REGEX = void 0;
/**
 * Regex matching legal names for POD entries.  This is intended to make
 * entry names usable as identifiers in TypeScript and other languages.
 */
exports.POD_NAME_REGEX = new RegExp(/^[A-Za-z_]\w*$/);
/**
 * Identifier for EdDSA public key string type.
 */
exports.EDDSA_PUBKEY_TYPE_STRING = "eddsa_pubkey";
/**
 * Regex matching legal values for types encoded as strings by
 * {@link podValueToRawValue}. This matches strings of the form
 * `pod_${PODName}:${string}`.
 */
exports.POD_STRING_TYPE_REGEX = new RegExp(/pod_([A-Za-z_]\w*):(.*)$/);
/**
 * Minimum legal value of a `cryptographic` entry value.
 */
exports.POD_CRYPTOGRAPHIC_MIN = 0n;
/**
 * Maximum legal value of a `cryptographic` entry value.
 */
exports.POD_CRYPTOGRAPHIC_MAX = 
// Prime order of the alt_bn128 curve.
// Should always be equal to the `r` constant in @zk-kit/baby-jubjub.
21888242871839275222246405745257275088548364400416034343698204186575808495617n -
    1n;
/**
 * Minimum legal value of an `int` entry value.
 */
exports.POD_INT_MIN = -(1n << 63n);
/**
 * Maximum legal value of an `int` entry value.
 */
exports.POD_INT_MAX = (1n << 63n) - 1n;
/**
 * Type constructor for EdDSA (Baby Jubjub) public keys.
 */
function PODEdDSAPublicKeyValue(value) {
    return { type: exports.EDDSA_PUBKEY_TYPE_STRING, value };
}
exports.PODEdDSAPublicKeyValue = PODEdDSAPublicKeyValue;
/**
 * Minimum legal value of a `date` entry value.
 */
exports.POD_DATE_MIN = new Date(-8640000000000000);
/**
 * Maximum legal value of a `date` entry value.
 */
exports.POD_DATE_MAX = new Date(8640000000000000);
/**
 * Convenience constant for a null PODValue.  Note that since users are not
 * required to use this constant, PODNullValues should not be compared using
 * `===`, but by examining the type or value fields.
 */
exports.PODNull = { type: "null", value: null };
/**
 * Fixed hash value representing a PODNullValue in a circuit.
 */
exports.POD_NULL_HASH = 0x1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1dn;
