// @pcd/semaphore-identity-pcd/dist/cjs/src/SemaphoreIdentityPCDPackage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SemaphoreIdentityPCDPackage = exports.getDisplayOptions = exports.deserialize = exports.serialize = exports.verify = exports.prove = void 0;
const semaphore_identity_v3_wrapper_1 = require("@pcd/semaphore-identity-v3-wrapper");
const util_1 = require("@pcd/util");
const json_bigint_1 = __importDefault(require("json-bigint"));
const uuid_1 = require("uuid");
const SemaphoreIdentityPCD_1 = require("./SemaphoreIdentityPCD");
const v4IdentityUtils_1 = require("./v4IdentityUtils");
async function prove(args) {
    return new SemaphoreIdentityPCD_1.SemaphoreIdentityPCD((0, uuid_1.v4)(), {
        identityV3: args.identityV3,
        identityV4: (0, v4IdentityUtils_1.v3tov4Identity)(args.identityV3)
    });
}
exports.prove = prove;
async function verify(pcd) {
    return pcd?.claim?.identityV3 !== undefined;
}
exports.verify = verify;
async function serialize(pcd) {
    return {
        type: SemaphoreIdentityPCD_1.SemaphoreIdentityPCDTypeName,
        pcd: json_bigint_1.default.stringify({
            type: pcd.type,
            id: pcd.id,
            // note this is the v3 identity, but is not named
            // 'identityV3' in the serialized format for backwards
            // compatibility
            identity: pcd.claim.identityV3.toString()
        })
    };
}
exports.serialize = serialize;
async function deserialize(serialized) {
    const { id, identity } = json_bigint_1.default.parse(serialized);
    (0, util_1.requireDefinedParameter)(id, "id");
    (0, util_1.requireDefinedParameter)(identity, "identity");
    const v3Identity = new semaphore_identity_v3_wrapper_1.Identity(identity);
    return new SemaphoreIdentityPCD_1.SemaphoreIdentityPCD(id, {
        identityV3: v3Identity,
        identityV4: (0, v4IdentityUtils_1.v3tov4Identity)(v3Identity)
    });
}
exports.deserialize = deserialize;
function getDisplayOptions(pcd) {
    return {
        header: "Semaphore Identity",
        displayName: "semaphore-id-" +
            pcd.claim.identityV3.commitment.toString().substring(0, 8)
    };
}
exports.getDisplayOptions = getDisplayOptions;
/**
 * PCD-conforming wrapper for the Semaphore zero-knowledge protocol. You can
 * find documentation of Semaphore here: https://semaphore.appliedzkp.org/docs/introduction
 */
exports.SemaphoreIdentityPCDPackage = {
    name: SemaphoreIdentityPCD_1.SemaphoreIdentityPCDTypeName,
    getDisplayOptions,
    prove,
    verify,
    serialize,
    deserialize
};
