// @pcd/passport-interface/dist/cjs/src/PassportInterface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PayloadType = exports.postPendingPCDMessage = exports.postSerializedMultiPCDMessage = exports.postSerializedPCDMessage = exports.constructZupassPcdProveAndAddRequestUrl = exports.constructZupassPcdMintRequestUrl = exports.constructZupassPcdAddRequestUrl = exports.constructZupassPcdGetRequestUrl = exports.getWithoutProvingUrl = exports.PCDRequestType = void 0;
var PCDRequestType;
(function (PCDRequestType) {
    PCDRequestType["Get"] = "Get";
    PCDRequestType["GetWithoutProving"] = "GetWithoutProving";
    PCDRequestType["Add"] = "Add";
    PCDRequestType["ProveAndAdd"] = "ProveAndAdd";
})(PCDRequestType || (exports.PCDRequestType = PCDRequestType = {}));
function getWithoutProvingUrl(zupassClientUrl, returnUrl, pcdType, postMessage = false) {
    const req = {
        type: PCDRequestType.GetWithoutProving,
        pcdType,
        returnUrl,
        postMessage
    };
    const encReq = encodeURIComponent(JSON.stringify(req));
    return `${zupassClientUrl}#/get-without-proving?request=${encReq}`;
}
exports.getWithoutProvingUrl = getWithoutProvingUrl;
function constructZupassPcdGetRequestUrl(zupassClientUrl, returnUrl, pcdType, args, options, postMessage = false) {
    const req = {
        type: PCDRequestType.Get,
        returnUrl: returnUrl,
        args: args,
        pcdType,
        options,
        postMessage
    };
    const encReq = encodeURIComponent(JSON.stringify(req));
    return `${zupassClientUrl}#/prove?request=${encReq}`;
}
exports.constructZupassPcdGetRequestUrl = constructZupassPcdGetRequestUrl;
function constructZupassPcdAddRequestUrl(zupassClientUrl, returnUrl, pcd, folder, postMessage = false, redirectToFolder) {
    const req = {
        type: PCDRequestType.Add,
        returnUrl,
        pcd,
        folder,
        postMessage,
        redirectToFolder
    };
    const eqReq = encodeURIComponent(JSON.stringify(req));
    return `${zupassClientUrl}#/add?request=${eqReq}`;
}
exports.constructZupassPcdAddRequestUrl = constructZupassPcdAddRequestUrl;
function constructZupassPcdMintRequestUrl(zupassClientUrl, mintUrl, returnUrl, pcd, folder, postMessage = false, redirectToFolder) {
    const req = {
        type: PCDRequestType.Add,
        mintUrl,
        returnUrl,
        pcd,
        folder,
        postMessage,
        redirectToFolder
    };
    const eqReq = encodeURIComponent(JSON.stringify(req));
    return `${zupassClientUrl}#/add?request=${eqReq}`;
}
exports.constructZupassPcdMintRequestUrl = constructZupassPcdMintRequestUrl;
function constructZupassPcdProveAndAddRequestUrl(zupassClientUrl, returnUrl, pcdType, args, options, returnPCD, folder, postMessage = false) {
    const req = {
        type: PCDRequestType.ProveAndAdd,
        returnUrl: returnUrl,
        pcdType,
        args,
        options,
        returnPCD,
        folder,
        postMessage
    };
    const eqReq = encodeURIComponent(JSON.stringify(req));
    return `${zupassClientUrl}#/add?request=${eqReq}`;
}
exports.constructZupassPcdProveAndAddRequestUrl = constructZupassPcdProveAndAddRequestUrl;
function postSerializedPCDMessage(window, serialized) {
    window.postMessage({ encodedPCD: JSON.stringify(serialized) }, "*");
}
exports.postSerializedPCDMessage = postSerializedPCDMessage;
function postSerializedMultiPCDMessage(window, pcds) {
    window.postMessage({ multiplePCDs: JSON.stringify(pcds) }, "*");
}
exports.postSerializedMultiPCDMessage = postSerializedMultiPCDMessage;
function postPendingPCDMessage(window, pending) {
    window.postMessage({ encodedPendingPCD: JSON.stringify(pending) }, "*");
}
exports.postPendingPCDMessage = postPendingPCDMessage;
var PayloadType;
(function (PayloadType) {
    PayloadType["RedirectTopicData"] = "topic-data";
    PayloadType["NullifierHash"] = "nullifier-hash";
    PayloadType["AnonTopicDataPayload"] = "anon-topic-data-payload";
    PayloadType["ReactData"] = "react-data";
})(PayloadType || (exports.PayloadType = PayloadType = {}));
